### PR TITLE
[POC] Add Spring Social support for Microsoft Graph

### PR DESCRIPTION
**Pull request by ZhijunZhao**
@yungez @xscript @aovechki 

As discussed, I'll add API binding for email reading scenario.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -48,23 +49,29 @@
...
-        <azure.doucmentdb.autoconfigure.version>${project.version}</azure.doucmentdb.autoconfigure.version>
-        <spring.data.documentdb.version>${project.version}</spring.data.documentdb.version>
-        <azure.storage.starter.version>${project.version}</azure.storage.starter.version>
-        <azure.storage.autoconfigure.version>${project.version}</azure.storage.autoconfigure.version>
-        <azure.servicebus.autoconfigure.version>${project.version}</azure.servicebus.autoconfigure.version>
-        <azure.servicebus.starter.version>${project.version}</azure.servicebus.starter.version>
-        <azure.mediaservices.starter.version>${project.version}</azure.mediaservices.starter.version>
-        <azure.mediaservices.autoconfigure.version>${project.version}</azure.mediaservices.autoconfigure.version>
-        <azure.spring.common.version>${project.version}</azure.spring.common.version>
+        <azure.doucmentdb.starter.version>${bom.version}</azure.doucmentdb.starter.version>
```
Comment (ZhijunZhao) [inline pom.xml]: @rajadilipkolli The project.version property will be evaluated to user's module version, whose version should be independent of our starter version. So, I have to modify your change to the right. The sample projects are examples that user can specify whatever the version they want, not necessarily 0.1.4.

Comment (rajadilipkolli) [inline pom.xml]: Agreed

File: microsoft-graph-spring-social-starter-sample/src/main/resources/templates/connect/microsoftConnect.html [hunk 2]
```diff
@@ -0,0 +1,18 @@
+<html>
+<head>
+    <title>Hello MS Graph</title>
```
Comment (aovechki) [inline microsoft-graph-spring-social-starter-sample/src/main/resources/templates/connect/microsoftConnect.html]: Let's not use abbreviation here and the rest of the changes

File: pom.xml (hunk 1 above)
Comment (aovechki) [inline pom.xml]: Please correct the typo

File: spring-social-microsoft-graph/src/test/resources/user.json [hunk 3]
```diff
@@ -0,0 +1,15 @@
+{
+  "id": "16f5a7b6-5a15-4568-aa5a-31bb117e9967",
+  "businessPhones": [
+    "123"
+  ],
+  "displayName": "Anne Weiler",
+  "givenName": "Anne",
+  "jobTitle": "Manufacturing Lead",
+  "mail": "annew@CIE493742.onmicrosoft.com",
+  "mobilePhone": "+1 3528700812",
```
Comment (aovechki) [inline spring-social-microsoft-graph/src/test/resources/user.json]: This looks like real data - please change to fictitious

Comment (ZhijunZhao) [inline spring-social-microsoft-graph/src/test/resources/user.json]: I think it's fictitious already. The data is from the Microsoft Graph Explorer. 

Click the `GET my profile` in the `Getting Started` tab, without logging in. 
https://developer.microsoft.com/en-us/graph/graph-explorer#

File: spring-social-microsoft-graph/src/test/java/com/microsoft/azure/msgraph/api/impl/UserTemplateTest.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    @Test
+    public void getUserProfileValid() throws Exception {
+        mockServer
+                .expect(requestTo(microsoft.getGraphAPI("me")))
+                .andExpect(method(GET))
+                .andExpect(header("Authorization", "Bearer access token"))
+                .andRespond(withSuccess(jsonResource("user"), MediaType.APPLICATION_JSON));
+
+        final User user = microsoft.userOperations().getUserProfile();
+        assertThat(user.getDisplayName()).isEqualTo("Anne Weiler");
```
Comment (ZhijunZhao) [inline spring-social-microsoft-graph/src/test/java/com/microsoft/azure/msgraph/api/impl/UserTemplateTest.java]: It's from test/resources/user.json

File: spring-social-microsoft-graph/src/main/java/com/microsoft/azure/msgraph/connect/MicrosoftAdapter.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+import com.microsoft.azure.msgraph.api.User;
+import org.springframework.social.ApiException;
+import org.springframework.social.connect.ApiAdapter;
+import org.springframework.social.connect.ConnectionValues;
+import org.springframework.social.connect.UserProfile;
+import org.springframework.social.connect.UserProfileBuilder;
+
+public class MicrosoftAdapter implements ApiAdapter<Microsoft> {
+    @Override
+    public boolean test(Microsoft microsoft) {
```
Comment (ZhijunZhao) [inline spring-social-microsoft-graph/src/main/java/com/microsoft/azure/msgraph/connect/MicrosoftAdapter.java]: It's an interface from base class ApiAdapter.

File: spring-social-microsoft-graph/pom.xml [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        <artifactId>azure-spring-boot-starter-parent</artifactId>
+        <version>0.1.4</version>
+        <relativePath>../azure-spring-boot-starter-parent/pom.xml</relativePath>
+    </parent>
+
+    <artifactId>spring-social-microsoft-graph</artifactId>
+    <packaging>jar</packaging>
+
+    <name>Spring Social Microsoft Graph</name>
+    <description>extension for Spring Social to enable connectivity with Microsoft Graph and an API binding for
```
Comment (ZhijunZhao) [inline spring-social-microsoft-graph/pom.xml]: Got it! We'll finalize the name before releasing it.

File: microsoft-graph-spring-social-starter-sample/README.md [hunk 7]
```diff
@@ -0,0 +1,16 @@
+## Usage
+
+### Register your application
+
+Go to [Application Registration Portal](https://apps.dev.microsoft.com/#/appList) and register your app.
```
Comment (ZhijunZhao) [inline microsoft-graph-spring-social-starter-sample/README.md]: yes, I logged in with my outlook.com account.

File: microsoft-graph-spring-social-starter-sample/src/main/resources/templates/connect/microsoftConnect.html [hunk 8]
```diff
@@ -0,0 +1,18 @@
+<html>
+<head>
+    <title>Hello Microsoft Graph</title>
+</head>
+<body>
+<h3>Connect to MS Graph</h3>
```
Comment (yungez) [inline microsoft-graph-spring-social-starter-sample/src/main/resources/templates/connect/microsoftConnect.html]: no abbreviation for all...

File: microsoft-graph-spring-social-starter/pom.xml [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+        <artifactId>azure-spring-boot-starter-parent</artifactId>
+        <groupId>com.microsoft.azure</groupId>
+        <version>0.1.4</version>
+        <relativePath>../azure-spring-boot-starter-parent/pom.xml</relativePath>
+    </parent>
+
+    <artifactId>microsoft-graph-spring-social-starter</artifactId>
+    <packaging>jar</packaging>
+
+    <name>Microsoft Graph Spring Social Starter</name>
```
Comment (yungez) [inline microsoft-graph-spring-social-starter/pom.xml]: I think it's better to have a readme under starter folder to explain what's this starter for, and basic functionality. although professional user could guess some point from naming. it's still necessary to have such readme.

**Pull request merged**